GTMAppAuthSwift/Sources/GTMOAuth2KeychainCompatibility.swift

```diff
@@ -0,0 +1,321 @@
+    for name: String,
+    clientID: String,
+    clientSecret: String
+  ) throws -> GTMAppAuthFetcherAuthorization {
+    return try authorizeFromKeychain(
+      forName: name,
+      tokenURL: googleTokenURL,
+      redirectURI: nativeClientRedirectURI,
+      clientID: clientID,
+      clientSecret: clientSecret
+    )
+  }
+
+/// Saves the authorization state to the keychain, in a GTMOAuth2 compatible manner.
+///
+/// - Parameters:
+///   - authorization: The `GTMAppAuthFetcherAuthorization` to save to the keychain.
+///   - name: The keychain name.
+/// - Throws: An Instance of `GTMOAuth2KeychainCompatibility.Error` arising from the save.
+  @available(*, deprecated, message: "Use GTMAppAuthFetcherAuthorization.save(authorization:with:)")
+  @objc public static func save(
+    authorization: GTMAppAuthFetcherAuthorization,
+    for name: String
+  ) throws {
+    guard let password = persistenceResponseStringForAuthorization(authorization) else {
+      throw Error.failedToCreateResponseStringFromAuthorization(authorization)
+    }
+
+    let keychain = keychain ?? GTMKeychain()
+    try keychain.save(password: password, forName: name)
+  }
+
+/// Removes stored tokens, such as when the user signs out.
+///
+/// - Parameters:
+///   - name: The keychain name.
+/// - Throws: An Instance of `GTMOAuth2KeychainCompatibility.Error` arising from the removal.
+  @objc public static func removeAuthorizationFromKeychain(for name: String) throws {
+    let keychain = keychain ?? GTMKeychain()
+
+    try keychain.removePasswordFromKeychain(forName: name)
+  }
+
+  // MARK: - OAuth2 Utilities
+
+  private static var googleAuthorizationURL: URL {
+    return URL(string: "https://accounts.google.com/o/oauth2/v2/auth")!
+  }
+
+  private static var googleTokenURL: URL {
+    return URL(string: "https://www.googleapis.com/oauth2/v4/token")!
+  }
+
+  private static var googleRevocationURL: URL {
+    return URL(string: "https://accounts.google.com/o/oauth2/revoke")!
+  }
+
+  private static var googleUserInfoURL: URL {
+    return URL(string: "https://www.googleapis.com/oauth2/v3/userinfo")!
+  }
+
+  private static var nativeClientRedirectURI: String {
+    return oobString
+  }
+
+  private static func dictionary(from keychainPassword: String) -> [String: String] {
+    let keyValueTuples: [(String, String)] = keychainPassword
+      .components(separatedBy: "&")
+      .compactMap {
+        let equalComps = $0.components(separatedBy: "=")
+        guard let key = equalComps.first,
+              let percentsRemovedKey = key.removingPercentEncoding,
+              let value = equalComps.last,
+              let percentsRemovedValue = value.removingPercentEncoding else {
+          return nil
+        }
+        return (percentsRemovedKey, percentsRemovedValue)
+      }
+    let passwordDictionary = Dictionary(uniqueKeysWithValues: keyValueTuples)
+    return passwordDictionary
+  }
+}
+
+public extension GTMOAuth2KeychainCompatibility {
```

Comment:
What's the reasoning for defining these elements in an extension vs in the implementation above?

Reply:
It's a common pattern in organizing code in Swift to make things more readable. Extensions like this are used when the code can be grouped in one place and is used across multiple "units." These units could be different places of the same source file, and perhaps event outside of it.

go/swift-style#extensions

Reply:
Is there a reuse opportunity here, or is improved readability the goal?

Reply:
It's both. "Used across multiple units" refers to the nested `Error` being used in this file and elsewhere, though that's somewhat tenuous here. I think using the extension here makes the code more readable in this case. Nesting the error under `GTMOAuth2KeychainCompatibility` is nice because it scopes the error to that name, but defining the error within the rest of the implementation would make it harder to read everything and harder to find `GTMOAuth2KeychainCompatibility.Error` when debugging (say, during development when something is throwing an error and you want to understand why).

Reply:
I think the style guide is using "units" to refer to related implementation elements that could be grouped into extensions for improved readability.

I realized we've already used the pattern of defining the error type in an extension, so I'm not overly concerned about this particular case.  We can continue the discussion of decomposing via extension offline.